draft/XXXX-orm/README.md

```diff
@@ -0,0 +1,103 @@
+
+Furthermore, due to the flat use of stateless functions, performance becomes an issue, as the same few functions are called repeatedly by multiple functions, as they share no state. The temporary solution to this was caching, but by introducing an ORM, much of the need for this caching will go away, as this state can be preserved in a class instance.
+
+## Proposals
+
+### Classes
+The following subsections describe the classes that make up the Multinet ORM. Type annotations are included to more accurately describe the connections between the classes. If a type annotation is left out, it is either due to uncertainty, or because that method returns nothing useful.
+
+
+##### User
+```
+# Previously used the term "cookie"
+def get_session() -> str: ...
+def set_session(): ...
+def delete_session(): ...
+
+# Serializes the object to JSON
+def serialize() -> Dict: ...
+
+# Static
+def exists(user: UserInfo) -> bool: ...
+def register(user: UserInfo) -> User: ...
+def from_session(session_id: str) -> User: ...
+```
+
+##### Table
+```
+def rows() -> Iterable[Dict]: ...
+def row_count() -> int: ...
+def keys() -> Iterable[str]: ...
+def edge() -> bool: ...
+```
+
+##### Graph
+```
+def nodes() -> Iterable[Dict]: ...
+def node_tables() -> Iterable[Table]: ...
+def edge_table() -> Table: ...
+```
+
+
+##### Workspace
+The workspace class would contain the following methods:
+```
+
+def table(name: str) -> Table: ...
+def tables() -> Iterable[Table]: ...
+
+def graph(name: str) -> Graph: ...
+def graphs() -> Iterable[Graph]: ...
+
+def metadata() -> Dict: ...
+
+def rename(new_name: str): ...
+def delete(): ...
+
+def create_table(name: str, rows: Iterable[Dict], edge: bool = False) -> Table: ...
+def create_aql_table(name: str, aql: str) -> Table: ...
+def delete_table(name: str): ...
+
+def create_graph(
+  name: str,
+  edge_table: str,
+  from_collections: Iterable[str],
+  to_collections: Iterable[str],
+) -> Graph: ...
+def delete_graph(name: str): ...
+
+def permissions() -> Dict: ...
+
+# Static
+def create(name: str) -> Workspace: ...
+def exists(name: str) -> bool: ...
+```
+
+
+### Open Questions
+
+* Should nodes be included in the ORM, or should they remain as dictionaries?
+* Should `create`, `delete`, etc. be attached to the classes they are operating on? This was the original proposal, but after discussion was decided against. The reason it's mentioned here is to highlight the benefit of such a model, mainly:
+  * Methods are more tightly coupled to the classes they are affecting
+  * Methods are spread more evenly between classes, instead of `Workspace` containing the bulk of the operations
+
+  However, the main reason against this approach is that it involves instantiation of objects possibly before they exist, which may be confusing.
```

Comment:
This is more evidence that the methods should stay with `Workspace`. Just to drive the main point home: we are modeling the manipulation of Arango collections from Arango databases; in that context, there's no notion of creating a collection without first being in the context of a workspace. So to shift over to Multinet's idea of "workspaces" and "tables", to introduce this idea of tables preceding their workspaces is indeed confusing, and just isn't needed in order to make this work.

Reply:
This is a good point. I was already convinced on your methodology for distributing methods, but this finally made it clear as day to me. That is, in my alternative approach, you would be able to, for example, run `Table(name).create()` without ever specifying a workspace, which is invalid.